File: frontend/src/types/video.ts
export type VideoStatus = 'uploaded' | 'processing' | 'completed' | 'failed';

export interface Video {
  id: string;
  filename: string;
  original_filename: string;
  status: VideoStatus;
  has_voice_file: boolean;
  duration_seconds: number | null;
  total_frames_analyzed: number | null;
  unique_entities_count: number | null;
  created_at: string;
  processed_at: string | null;
}

export interface VideoDetail extends Video {
  video_path: string;
  voice_path: string | null;
  interval_seconds: number;
  current_stage: string | null;
  progress_percentage: number | null;
  progress_message: string | null;
  report_path: string | null;
  frames_directory: string | null;
  error_message: string | null;
  updated_at: string;
}

export interface VideoStatusResponse {
  video_id: string;
  status: VideoStatus;
  progress_percentage: number | null;
  current_stage: string | null;
  progress_message: string | null;
  error_message: string | null;
}

export interface VideoReport {
  video_id: string;
  filename: string;
  duration_seconds: number;
  total_frames_analyzed: number;
  interval_seconds: number;
  unique_entities: string[];
  entity_appearances: Record<string, any>;
  timeline: any[];
  statistics: Record<string, any>;
  consolidated_timeline: Array<{
    entity: string;
    start: string;
    end: string;
    start_second: number;
    end_second: number;
    duration_seconds: number;
  }>;
}

export interface VideoListResponse {
  videos: Video[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface UploadResponse {
  video_id: string;
  filename: string;
  status: string;
  has_voice_file: boolean;
  interval_seconds: number;
  message: string;
}

File: api/schemas/video.py
"""
Video-related schemas for request and response validation.
"""
from typing import Optional, List, Dict, Any
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field, field_validator


class VideoUploadResponse(BaseModel):
    """Response after uploading a video."""
    video_id: str
    filename: str
    status: str
    has_voice_file: bool
    interval_seconds: int
    message: str

    @field_validator('video_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        """Convert UUID to string."""
        return str(v) if isinstance(v, UUID) else v


class VideoStatus(BaseModel):
    """Processing status information."""
    video_id: str
    status: str
    progress_percentage: Optional[float] = None
    current_stage: Optional[str] = None
    progress_message: Optional[str] = None
    error_message: Optional[str] = None

    @field_validator('video_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        """Convert UUID to string."""
        return str(v) if isinstance(v, UUID) else v


class VideoDetail(BaseModel):
    """Detailed video information."""
    id: str
    filename: str
    original_filename: str
    video_path: str
    voice_path: Optional[str] = None
    interval_seconds: int
    has_voice_file: bool
    status: str
    current_stage: Optional[str] = None
    progress_percentage: Optional[float] = None
    progress_message: Optional[str] = None
    report_path: Optional[str] = None
    frames_directory: Optional[str] = None
    duration_seconds: Optional[float] = None
    total_frames_analyzed: Optional[int] = None
    unique_entities_count: Optional[int] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    processed_at: Optional[datetime] = None

    @field_validator('id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        """Convert UUID to string."""
        return str(v) if isinstance(v, UUID) else v

    class Config:
        from_attributes = True


class VideoSummary(BaseModel):
    """Summary video information for list views."""
    id: str
    filename: str
    status: str
    has_voice_file: bool
    duration_seconds: Optional[float] = None
    total_frames_analyzed: Optional[int] = None
    unique_entities_count: Optional[int] = None
    created_at: datetime
    processed_at: Optional[datetime] = None

    @field_validator('id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        """Convert UUID to string."""
        return str(v) if isinstance(v, UUID) else v

    class Config:
        from_attributes = True


class VideoList(BaseModel):
    """Paginated list of videos."""
    videos: List[VideoSummary]
    total: int
    page: int
    limit: int
    total_pages: int


class FrameInfo(BaseModel):
    """Information about a single frame."""
    frame_number: int
    timestamp: float
    frame_path: str
    entities_detected: List[str] = []


class VideoReport(BaseModel):
    """Video processing report."""
    video_id: str
    filename: str
    duration_seconds: float
    total_frames_analyzed: int
    interval_seconds: int
    unique_entities: List[str]
    entity_appearances: Dict[str, Any]
    timeline: List[Dict[str, Any]]
    statistics: Dict[str, Any]

    @field_validator('video_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        """Convert UUID to string."""
        return str(v) if isinstance(v, UUID) else v


class ProcessVideoRequest(BaseModel):
    """Request to process a video."""
    interval_seconds: Optional[int] = Field(default=5, ge=1, le=60)


class DeleteVideoResponse(BaseModel):
    """Response after deleting a video."""
    success: bool
    message: str
    video_id: str

    @field_validator('video_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        """Convert UUID to string."""
        return str(v) if isinstance(v, UUID) else v
